bitwarden_license/src/Commercial.Infrastructure.EntityFramework/Repositories/AccessPolicyRepository.cs

```diff
@@ -0,0 +1,201 @@
+        {
+            var dbContext = GetDatabaseContext(scope);
+            var entity = await dbContext.AccessPolicies.FindAsync(id);
+
+            if (entity == null)
+            {
+                return null;
+            }
+
+            switch (entity)
+            {
+                case UserProjectAccessPolicy accessPolicy:
+                    await dbContext.Entry(accessPolicy)
+                        .Reference(e => e.OrganizationUser)
+                        .Query().Include(e => e.User)
+                        .LoadAsync();
+                    return Mapper.Map<Core.Entities.UserProjectAccessPolicy>(accessPolicy);
+                case GroupProjectAccessPolicy accessPolicy:
+                    await dbContext.Entry(accessPolicy)
+                        .Reference(e => e.Group)
+                        .LoadAsync();
+                    return Mapper.Map<Core.Entities.GroupProjectAccessPolicy>(accessPolicy);
+                case ServiceAccountProjectAccessPolicy accessPolicy:
+                    await dbContext.Entry(accessPolicy)
+                        .Reference(e => e.ServiceAccount)
+                        .LoadAsync();
+                    return Mapper.Map<Core.Entities.ServiceAccountProjectAccessPolicy>(accessPolicy);
+                default:
+                    throw new ArgumentException("Unsupported access policy type");
+            }
+        }
+    }
+
+    public async Task ReplaceAsync(Core.Entities.BaseAccessPolicy baseAccessPolicy)
+    {
+        using (var scope = ServiceScopeFactory.CreateScope())
+        {
+            var dbContext = GetDatabaseContext(scope);
+            var entity = await dbContext.AccessPolicies.Where(c => c.Id == baseAccessPolicy.Id).FirstOrDefaultAsync();
+            if (entity != null)
+            {
+                dbContext.AccessPolicies.Attach(entity);
+                entity.Write = baseAccessPolicy.Write;
+                entity.Read = baseAccessPolicy.Read;
+                await dbContext.SaveChangesAsync();
+            }
+        }
+    }
+
+    public async Task<List<Core.Entities.BaseAccessPolicy>?> GetManyByProjectId(Guid id)
+    {
+        using (var scope = ServiceScopeFactory.CreateScope())
+        {
+            var dbContext = GetDatabaseContext(scope);
+
+            var userEntities = await dbContext.UserProjectAccessPolicy
+                    .Where(c => c.GrantedProjectId == id)
+                    .Include(e => e.OrganizationUser.User)
+                    .ToListAsync();
+            var groupEntities = await dbContext.GroupProjectAccessPolicy
+                    .Where(c => c.GrantedProjectId == id)
+                    .Include(e => e.Group)
+                    .ToListAsync();
+            var serviceAccountEntities = await dbContext.ServiceAccountProjectAccessPolicy
+                    .Where(c => c.GrantedProjectId == id)
+                    .Include(e => e.ServiceAccount)
+                    .ToListAsync();
```

Comment:
This kind of defeats the purpose of EF TPH since we need to run 3 queries instead of 1?

Reply:
I think this should work. I haven't inspected the query (since the tests uses a memory one which doesn't provide a string based query) but my assumption is it would use the deliminator in the queries.

```csharp
dbContext.AccessPolicies.Where(ap =>
    ((UserProjectAccessPolicy)ap).GrantedProjectId == id ||
    ((GroupProjectAccessPolicy)ap).GrantedProjectId == id ||
    ((ServiceAccountProjectAccessPolicy)ap).GrantedProjectId == id);
```

It would then return the correct classes i.e. `ServiceAccountProjectAccessPolicy` which can be checked when cast back to the core entities.